File: 2020/05_BinaryBoarding/phone.lua
-- ======================================================================
-- Binary Boarding
--   Advent of Code 2020 Day 05 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         p h o n e . l u a
-- ======================================================================
-- A solver for the Advent of Code 2020 Day 05 puzzle

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local bpass = require('bpass')

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Phone = { part2 = false, text = {}, passes = {} }

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                  Phone
-- ======================================================================

function Phone:Phone (o)
  -- Object for Binary Boarding 
  
  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.passes = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Phone:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do

    -- 2. Add the number to the entries
    local pass = bpass:Bpass({part2=self.part2, text=line})
    self.passes[pass.seat] = pass
    end

end

function Phone:highest_seat()
  -- Return the ID of the highest seat
  
  -- 1. Start with nothing
  local result = 0
  
  -- 2. Loop for all the boarding passes
  for seat, _ in pairs(self.passes) do
    
    -- 3. If this pass has a higher seat number, save it
    if seat > result then
      result = seat
    end
  end
  
  -- 4. Return the highest seat number
  return result
end

function Phone:lowest_seat()
  -- Return the ID of the lowest seat
  
  -- 1. Start with a very high seat number indeed
  local result = 99999
  
  -- 2. Loop for all the boarding passes
  for seat, _ in pairs(self.passes) do
    
    -- 3. If this pass has a lower seat number, save it
    if seat < result then
      result = seat
    end
  end
  
  -- 4. Return the highest seat number
  return result
end
  
function Phone:find_empty()
  -- Returns an empty seat number
  
  -- 1. Get the high and low seat numbers
  local high = self:highest_seat()
  local low = self:lowest_seat()
  
  -- 2. Loop for the seat numbers between them
  for seat = low, high do
    
    -- 3. If we have no record of this seat number, return it
    if nil == self.passes[seat] then
      return seat
    end
  end
  
  -- 4. Every seat is full
  return nil
end
  
  
function Phone:part_one(args)
  -- "Returns the solution for part one"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:highest_seat()
end


function Phone:part_two(args)
  -- "Returns the solution for part two"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  return self:find_empty()
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Phone

-- ======================================================================
-- end                      p h o n e . l u a                     end
-- ======================================================================

File: 2020/05_BinaryBoarding/phone.ts
// ======================================================================
// Binary Boarding
//   Advent of Code 2020 Day 05 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           p h o n e . t s
//
// A solver for the Advent of Code 2020 Day 05 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
import { Bpass } from './bpass';

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                  Phone
// ======================================================================

export class Phone {
  // Object for Binary Boarding
  text: string[];

  part2: boolean;

  passes: Bpass[];

  constructor(text: string[], part2 = false) {
    // Create a Phone object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.passes = [];

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      for (let indx = 0; indx < this.text.length; indx += 1) {
        this.passes.push(new Bpass(this.text[indx], this.part2));
      }
    }
  }

  highestSeat(): number {
    // Returns the number of the highest seat
    // 1. Start with nothing
    let result = -1;

    // 2. Loop for all of the boarding passes
    for (let indx = 0; indx < this.passes.length; indx += 1) {
      // 3. If the seat number of the boarding pass is higher, save the seat number
      if (this.passes[indx].seat > result) {
        result = this.passes[indx].seat;
      }
    }
    // 4. Return the higest seat number
    return result;
  }

  findSeat(): number {
    // Returns the number of the seat no one else has
    // 1. What is the highest seat number
    const highest = this.highestSeat();

    // 2. Work down to find your seat
    for (let seat = highest; seat > 0; seat -= 1) {
      let found = false;

      // 3. Loop for all of the boarding passes
      for (let indx = 0; indx < this.passes.length; indx += 1) {
        if (seat === this.passes[indx].seat) {
          found = true;
        }
      }

      // 4. If there is no boarding pass for this seat, it is ours
      if (!found) {
        return seat;
      }
    }

    // 5. That's odd
    return -1;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.findSeat();
    }
    return this.highestSeat();
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                         p h o n e . t s                        end
// ======================================================================
